src/NLog/Internal/MultiFileWatcher.cs

```diff
@@ -107,33 +136,33 @@ internal void Watch(string fileName)
                 return;
             }
 
-            var watcher = new FileSystemWatcher
+            lock (this)
```

Comment:
I think to keep this thread-safe, the `ContainsKey()` and `Add()` need to be enclosed in the same `lock` block, otherwise the same file might be added by two threads at the same time.